blk/components/virt.c

```diff
@@ -0,0 +1,344 @@
+                case WRITE_BLOCKS:
+                    blk_enqueue_resp(&h, SUCCESS, drv_success_count, cli_data.cli_req_id);
+                    break;
+                case FLUSH:
+                case BARRIER:
+                    blk_enqueue_resp(&h, SUCCESS, drv_success_count, cli_data.cli_req_id);
+                    break;
+            }
+        } else {
+            // When more error conditions are added, this will need to be updated to a switch statement
+            blk_enqueue_resp(&h, SEEK_ERROR, drv_success_count, cli_data.cli_req_id);
+        }
+        
+        // @ericc: Print out first 512bytes of data for debugging
+        // if (cli_data.cli_req_id == 0) {
+        //     LOG_BLK_VIRT("drv_addr: 0x%lx\n", cli_data.drv_addr);
+        //     char* addr = (char*)cli_data.drv_addr;
+
+        //     for (int i = 0; i < 512; i++) {
+        //         sddf_dprintf("%02x ", addr[i]);
+        //     }
+
+        //     sddf_dprintf("\n");
+        // }
+        
+        // Notify corresponding client
+        microkit_notify(clients[cli_data.cli_id].ch);
+    }
+}
+
+static void handle_client(int cli_id) {
+    blk_queue_handle_t h = clients[cli_id].queue_h;
+
+    blk_request_code_t cli_code;
+    uintptr_t cli_addr;
+    uint32_t cli_block_number;
+    uint16_t cli_count;
+    uint32_t cli_req_id;
+
+    uintptr_t drv_addr;
+    uint32_t drv_block_number;
+    uint64_t drv_req_id;
+    while (!blk_req_queue_empty(&h)) {
+        blk_dequeue_req(&h, &cli_code, &cli_addr, &cli_block_number, &cli_count, &cli_req_id);
+
+        if (cli_code == READ_BLOCKS || cli_code == WRITE_BLOCKS) {
+            // Check if client request is within its allocated bounds
+            if (cli_block_number + cli_count > clients[cli_id].sectors) {
+                blk_enqueue_resp(&h, SEEK_ERROR, 0, cli_req_id);
+                continue;
+            }
+        }
+
+        drv_block_number = cli_block_number + (clients[cli_id].start_sector / (BLK_TRANSFER_SIZE / MBR_SECTOR_SIZE));
+
+        switch(cli_code) {
+            case READ_BLOCKS:
+                if (blk_req_queue_full(&drv_h) || datastore_full(&ds) || fsmem_full(&fsmem_data, cli_count)) {
+                    continue;
+                }
+                // Allocate driver data buffers
+                fsmem_alloc(&fsmem_data, &drv_addr, cli_count);
+                break;
+            case WRITE_BLOCKS:
+                if (blk_req_queue_full(&drv_h) || datastore_full(&ds) || fsmem_full(&fsmem_data, cli_count)) {
+                    continue;
+                }
+                // Allocate driver data buffers
+                fsmem_alloc(&fsmem_data, &drv_addr, cli_count);
+                // Copy data buffers from client to driver
+                memcpy((void *)drv_addr, (void *)cli_addr, BLK_TRANSFER_SIZE * cli_count);
+                // Flush the cache
+                seL4_ARM_VSpace_Clean_Data(3, drv_addr, drv_addr + (BLK_TRANSFER_SIZE * cli_count));
```

Comment:
```suggestion
                cache_clean(drv_addr, drv_addr + (BLK_TRANSFER_SIZE * cli_count));
```

from `util/cache.c`. The cache clean operation, unlike invalidate, is available to user-space.